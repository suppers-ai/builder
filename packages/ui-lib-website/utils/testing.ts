// Testing utilities for component quality assurance

/**
 * Accessibility testing utilities
 */
export class AccessibilityTester {
  /**
   * Check if element has proper ARIA attributes
   */
  static checkAriaAttributes(element: Element): string[] {
    const issues: string[] = [];

    // Check for interactive elements without proper roles
    const interactiveElements = ["button", "a", "input", "select", "textarea"];
    if (interactiveElements.includes(element.tagName.toLowerCase())) {
      if (!element.getAttribute("role") && element.tagName.toLowerCase() === "div") {
        issues.push("Interactive div element missing role attribute");
      }
    }

    // Check for images without alt text
    if (element.tagName.toLowerCase() === "img") {
      if (!element.getAttribute("alt")) {
        issues.push("Image missing alt attribute");
      }
    }

    // Check for form inputs without labels
    if (["input", "select", "textarea"].includes(element.tagName.toLowerCase())) {
      const id = element.getAttribute("id");
      const ariaLabel = element.getAttribute("aria-label");
      const ariaLabelledBy = element.getAttribute("aria-labelledby");

      if (!ariaLabel && !ariaLabelledBy) {
        if (!id || !document.querySelector(`label[for="${id}"]`)) {
          issues.push("Form input missing accessible label");
        }
      }
    }

    // Check for proper heading hierarchy
    if (element.tagName.match(/^H[1-6]$/)) {
      const level = parseInt(element.tagName.charAt(1));
      const previousHeadings = Array.from(document.querySelectorAll("h1, h2, h3, h4, h5, h6"))
        .filter((h) =>
          document.compareDocumentPosition(element) & Node.DOCUMENT_POSITION_FOLLOWING
        );

      if (previousHeadings.length > 0) {
        const lastHeading = previousHeadings[previousHeadings.length - 1];
        const lastLevel = parseInt(lastHeading.tagName.charAt(1));

        if (level > lastLevel + 1) {
          issues.push(`Heading level ${level} skips hierarchy (previous was ${lastLevel})`);
        }
      }
    }

    return issues;
  }

  /**
   * Check color contrast ratios
   */
  static checkColorContrast(element: Element): string[] {
    const issues: string[] = [];
    const styles = globalThis.getComputedStyle(element);
    const color = styles.color;
    const backgroundColor = styles.backgroundColor;

    // Simple contrast check (would need more sophisticated algorithm in production)
    if (color && backgroundColor && color !== "transparent" && backgroundColor !== "transparent") {
      const colorLuminance = this.getLuminance(color);
      const bgLuminance = this.getLuminance(backgroundColor);
      const contrast = this.getContrastRatio(colorLuminance, bgLuminance);

      if (contrast < 4.5) {
        issues.push(`Low color contrast ratio: ${contrast.toFixed(2)} (minimum 4.5)`);
      }
    }

    return issues;
  }

  /**
   * Check keyboard navigation
   */
  static checkKeyboardNavigation(container: Element): string[] {
    const issues: string[] = [];
    const focusableElements = container.querySelectorAll(
      'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])',
    );

    focusableElements.forEach((element, index) => {
      const tabIndex = element.getAttribute("tabindex");

      // Check for positive tabindex (anti-pattern)
      if (tabIndex && parseInt(tabIndex) > 0) {
        issues.push(
          `Element has positive tabindex (${tabIndex}), which can disrupt natural tab order`,
        );
      }

      // Check for focus indicators
      const styles = globalThis.getComputedStyle(element, ":focus");
      if (!styles.outline || styles.outline === "none") {
        if (!styles.boxShadow && !styles.border) {
          issues.push("Element lacks visible focus indicator");
        }
      }
    });

    return issues;
  }

  private static getLuminance(color: string): number {
    // Simplified luminance calculation
    const rgb = color.match(/\d+/g);
    if (!rgb) return 0;

    const [r, g, b] = rgb.map((c) => {
      const channel = parseInt(c) / 255;
      return channel <= 0.03928 ? channel / 12.92 : Math.pow((channel + 0.055) / 1.055, 2.4);
    });

    return 0.2126 * r + 0.7152 * g + 0.0722 * b;
  }

  private static getContrastRatio(lum1: number, lum2: number): number {
    const lighter = Math.max(lum1, lum2);
    const darker = Math.min(lum1, lum2);
    return (lighter + 0.05) / (darker + 0.05);
  }
}

/**
 * Component testing utilities
 */
export class ComponentTester {
  /**
   * Test component rendering
   */
  static testRender(component: any, props: any = {}): boolean {
    try {
      // In a real implementation, this would use a testing framework
      // For now, just check if component can be instantiated
      const instance = component(props);
      return instance !== null && instance !== undefined;
    } catch (error) {
      console.error("Component render test failed:", error);
      return false;
    }
  }

  /**
   * Test component props
   */
  static testProps(
    component: any,
    requiredProps: string[],
    optionalProps: string[] = [],
  ): string[] {
    const issues: string[] = [];

    // Test with missing required props
    try {
      component({});
      if (requiredProps.length > 0) {
        issues.push("Component should require props but accepts empty object");
      }
    } catch (error) {
      // Expected for required props
    }

    // Test with valid props
    const validProps: any = {};
    requiredProps.forEach((prop) => {
      validProps[prop] = "test-value";
    });

    try {
      const result = component(validProps);
      if (!result) {
        issues.push("Component fails to render with valid props");
      }
    } catch (error) {
      issues.push(`Component throws error with valid props: ${error.message}`);
    }

    return issues;
  }

  /**
   * Test component accessibility
   */
  static testAccessibility(element: Element): string[] {
    const issues: string[] = [];

    issues.push(...AccessibilityTester.checkAriaAttributes(element));
    issues.push(...AccessibilityTester.checkColorContrast(element));
    issues.push(...AccessibilityTester.checkKeyboardNavigation(element));

    return issues;
  }
}

/**
 * Performance testing utilities
 */
export class PerformanceTester {
  /**
   * Measure component render time
   */
  static measureRenderTime(componentFn: () => void): number {
    const start = performance.now();
    componentFn();
    const end = performance.now();
    return end - start;
  }

  /**
   * Test memory usage (simplified)
   */
  static testMemoryUsage(testFn: () => void): any {
    if ("memory" in performance) {
      const beforeMemory = (performance as any).memory.usedJSHeapSize;
      testFn();
      const afterMemory = (performance as any).memory.usedJSHeapSize;

      return {
        before: beforeMemory,
        after: afterMemory,
        difference: afterMemory - beforeMemory,
      };
    }

    return { message: "Memory measurement not available" };
  }

  /**
   * Test for memory leaks in component cleanup
   */
  static testMemoryLeaks(
    createComponent: () => any,
    destroyComponent: (component: any) => void,
  ): boolean {
    const iterations = 100;
    const components: any[] = [];

    // Create many components
    for (let i = 0; i < iterations; i++) {
      components.push(createComponent());
    }

    // Destroy all components
    components.forEach((component) => destroyComponent(component));

    // Force garbage collection if available
    if ("gc" in window) {
      (window as any).gc();
    }

    // In a real test, you'd measure memory usage here
    return true; // Simplified
  }
}

/**
 * Cross-browser compatibility testing
 */
export class BrowserCompatibilityTester {
  /**
   * Check browser feature support
   */
  static checkFeatureSupport(): Record<string, boolean> {
    return {
      cssGrid: CSS.supports("display", "grid"),
      cssFlexbox: CSS.supports("display", "flex"),
      cssCustomProperties: CSS.supports("--test", "value"),
      intersectionObserver: "IntersectionObserver" in window,
      mutationObserver: "MutationObserver" in window,
      promiseSupport: "Promise" in window,
      fetchAPI: "fetch" in window,
      webComponents: "customElements" in window,
      serviceWorker: "serviceWorker" in navigator,
      localStorage: (() => {
        try {
          localStorage.setItem("test", "test");
          localStorage.removeItem("test");
          return true;
        } catch {
          return false;
        }
      })(),
    };
  }

  /**
   * Get browser information
   */
  static getBrowserInfo(): Record<string, string> {
    const ua = navigator.userAgent;

    return {
      userAgent: ua,
      platform: navigator.platform,
      language: navigator.language,
      cookieEnabled: navigator.cookieEnabled.toString(),
      onLine: navigator.onLine.toString(),
      vendor: navigator.vendor || "unknown",
      browserName: this.detectBrowser(ua),
      browserVersion: this.detectBrowserVersion(ua),
    };
  }

  private static detectBrowser(ua: string): string {
    if (ua.includes("Chrome")) return "Chrome";
    if (ua.includes("Firefox")) return "Firefox";
    if (ua.includes("Safari") && !ua.includes("Chrome")) return "Safari";
    if (ua.includes("Edge")) return "Edge";
    if (ua.includes("Opera")) return "Opera";
    return "Unknown";
  }

  private static detectBrowserVersion(ua: string): string {
    const match = ua.match(/(?:Chrome|Firefox|Safari|Edge|Opera)\/(\d+\.?\d*)/);
    return match ? match[1] : "Unknown";
  }
}

/**
 * Automated testing suite
 */
export class AutomatedTestSuite {
  private tests: Array<{
    name: string;
    test: () => Promise<boolean>;
    category: "accessibility" | "performance" | "compatibility" | "functionality";
  }> = [];

  /**
   * Add a test to the suite
   */
  addTest(
    name: string,
    test: () => Promise<boolean>,
    category: "accessibility" | "performance" | "compatibility" | "functionality",
  ) {
    this.tests.push({ name, test, category });
  }

  /**
   * Run all tests
   */
  async runTests(): Promise<{
    passed: number;
    failed: number;
    results: Array<{ name: string; passed: boolean; error?: string; category: string }>;
  }> {
    const results = [];
    let passed = 0;
    let failed = 0;

    for (const { name, test, category } of this.tests) {
      try {
        const result = await test();
        if (result) {
          passed++;
          results.push({ name, passed: true, category });
        } else {
          failed++;
          results.push({ name, passed: false, category });
        }
      } catch (error) {
        failed++;
        results.push({
          name,
          passed: false,
          error: error instanceof Error ? error.message : "Unknown error",
          category,
        });
      }
    }

    return { passed, failed, results };
  }

  /**
   * Generate test report
   */
  generateReport(testResults: any): string {
    const { passed, failed, results } = testResults;
    const total = passed + failed;
    const passRate = ((passed / total) * 100).toFixed(1);

    let report = `# Component Library Test Report\n\n`;
    report += `**Total Tests:** ${total}\n`;
    report += `**Passed:** ${passed}\n`;
    report += `**Failed:** ${failed}\n`;
    report += `**Pass Rate:** ${passRate}%\n\n`;

    // Group by category
    const categories = ["accessibility", "performance", "compatibility", "functionality"];

    categories.forEach((category) => {
      const categoryTests = results.filter((r: any) => r.category === category);
      if (categoryTests.length === 0) return;

      report += `## ${category.charAt(0).toUpperCase() + category.slice(1)} Tests\n\n`;

      categoryTests.forEach((result: any) => {
        const status = result.passed ? "✅" : "❌";
        report += `${status} **${result.name}**\n`;
        if (result.error) {
          report += `   Error: ${result.error}\n`;
        }
        report += "\n";
      });
    });

    return report;
  }
}
