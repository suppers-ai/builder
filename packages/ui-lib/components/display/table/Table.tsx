import { BaseComponentProps } from "../../types.ts";
import { ComponentChildren } from "preact";
import { useState } from "preact/hooks";
import { Pagination } from "../../navigation/pagination/Pagination.tsx";

export interface TableColumn {
  key: string;
  title: string;
  sortable?: boolean;
  render?: (value: unknown, row: unknown) => ComponentChildren;
  width?: string;
  align?: "left" | "center" | "right";
}

export interface TableProps extends BaseComponentProps {
  columns: TableColumn[];
  data: Record<string, unknown>[];
  zebra?: boolean;
  compact?: boolean;
  hover?: boolean;
  pinRows?: boolean;
  pinCols?: boolean;
  emptyMessage?: string;
  // Controlled mode props
  sortable?: boolean;
  sortColumn?: string;
  sortDirection?: "asc" | "desc";
  loading?: boolean;
  onSort?: (column: string, direction: "asc" | "desc") => void;
  // Pagination props
  paginated?: boolean;
  currentPage?: number;
  itemsPerPage?: number;
  totalItems?: number;
  onPageChange?: (page: number) => void;
  onItemsPerPageChange?: (itemsPerPage: number) => void;
  // Interactive features
  showControls?: boolean;
  allowItemsPerPageChange?: boolean;
  showStats?: boolean;
  itemsPerPageOptions?: number[];
}

export interface PaginatedTableProps extends BaseComponentProps {
  columns: TableColumn[];
  data: Record<string, unknown>[];
  // All table props
  zebra?: boolean;
  compact?: boolean;
  hover?: boolean;
  pinRows?: boolean;
  pinCols?: boolean;
  emptyMessage?: string;
  // Pagination configuration
  initialItemsPerPage?: number;
  itemsPerPageOptions?: number[];
  showStats?: boolean;
  showControls?: boolean;
  allowSorting?: boolean;
  allowItemsPerPageChange?: boolean;
}

export function Table({
  class: className = "",
  columns = [],
  data = [],
  zebra = false,
  compact = false,
  hover = false,
  pinRows = false,
  pinCols = false,
  emptyMessage = "No data available",
  sortable = false,
  sortColumn = "",
  sortDirection = "asc",
  loading = false,
  onSort,
  paginated = false,
  currentPage = 1,
  itemsPerPage = 10,
  totalItems,
  onPageChange,
  onItemsPerPageChange,
  showControls = false,
  allowItemsPerPageChange = false,
  showStats = false,
  itemsPerPageOptions = [5, 10, 20, 50],
  id,
  ...props
}: TableProps) {
  // Ensure columns and data are arrays
  const safeColumns = Array.isArray(columns) ? columns : [];
  const safeData = Array.isArray(data) ? data : [];

  // Calculate pagination values
  const totalPages = paginated ? Math.ceil((totalItems || safeData.length) / itemsPerPage) : 1;
  const startIndex = paginated ? (currentPage - 1) * itemsPerPage : 0;
  const endIndex = paginated ? startIndex + itemsPerPage : safeData.length;
  const displayData = paginated ? safeData.slice(startIndex, endIndex) : safeData;

  const tableClasses = [
    "table",
    zebra ? "table-zebra" : "",
    compact ? "table-xs" : "",
    hover ? "table-hover" : "",
    pinRows ? "table-pin-rows" : "",
    pinCols ? "table-pin-cols" : "",
    className,
  ].filter(Boolean).join(" ");

  // Controls component
  const renderControls = () => {
    if (!showControls) return null;

    return (
      <div class="flex flex-col sm:flex-row justify-between items-start sm:items-center gap-4 mb-4">
        {showStats && (
          <div class="stats stats-horizontal shadow">
            <div class="stat">
              <div class="stat-title">Total Items</div>
              <div class="stat-value text-primary">{totalItems || safeData.length}</div>
            </div>
            {paginated && (
              <>
                <div class="stat">
                  <div class="stat-title">Current Page</div>
                  <div class="stat-value text-secondary">{currentPage}</div>
                </div>
                <div class="stat">
                  <div class="stat-title">Total Pages</div>
                  <div class="stat-value text-accent">{totalPages}</div>
                </div>
              </>
            )}
          </div>
        )}

        {allowItemsPerPageChange && paginated && (
          <div class="form-control">
            <label class="label">
              <span class="label-text">Items per page</span>
            </label>
            <select
              class="select select-bordered select-sm"
              value={itemsPerPage}
              onChange={(e) =>
                onItemsPerPageChange?.(parseInt((e.target as HTMLSelectElement).value))}
            >
              {itemsPerPageOptions.map((option) => (
                <option key={option} value={option}>{option}</option>
              ))}
            </select>
          </div>
        )}
      </div>
    );
  };

  // Pagination info
  const renderPaginationInfo = () => {
    if (!paginated) return null;

    return (
      <div class="flex flex-col sm:flex-row justify-between items-center gap-4 mt-4">
        <div class="text-sm text-base-content/70">
          Showing {startIndex + 1}-{Math.min(endIndex, totalItems || safeData.length)} of{" "}
          {totalItems || safeData.length} items
        </div>

        <Pagination
          currentPage={currentPage}
          totalPages={totalPages}
          onPageChange={onPageChange}
          size="sm"
        />
      </div>
    );
  };

  const handleSort = (columnKey: string) => {
    if (!sortable || !onSort) return;

    const column = safeColumns.find((c) => c.key === columnKey);
    if (!column?.sortable) return;

    let newDirection: "asc" | "desc" = "asc";
    if (sortColumn === columnKey && sortDirection === "asc") {
      newDirection = "desc";
    }

    onSort(columnKey, newDirection);
  };

  const getSortIcon = (columnKey: string) => {
    if (!sortable || sortColumn !== columnKey) {
      return (
        <svg
          xmlns="http://www.w3.org/2000/svg"
          class="h-4 w-4 opacity-30"
          fill="none"
          viewBox="0 0 24 24"
          stroke="currentColor"
        >
          <path
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            d="M7 16V4m0 0L3 8m4-4l4 4m6 0v12m0 0l4-4m-4 4l-4-4"
          />
        </svg>
      );
    }

    return sortDirection === "asc"
      ? (
        <svg
          xmlns="http://www.w3.org/2000/svg"
          class="h-4 w-4"
          fill="none"
          viewBox="0 0 24 24"
          stroke="currentColor"
        >
          <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M5 15l7-7 7 7" />
        </svg>
      )
      : (
        <svg
          xmlns="http://www.w3.org/2000/svg"
          class="h-4 w-4"
          fill="none"
          viewBox="0 0 24 24"
          stroke="currentColor"
        >
          <path
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            d="M19 9l-7 7-7-7"
          />
        </svg>
      );
  };

  if (loading) {
    return (
      <div class="space-y-4">
        {renderControls()}
        <div class="overflow-x-auto">
          <table class={tableClasses} id={id} {...props}>
            <thead>
              <tr>
                {safeColumns.map((column) => (
                  <th
                    key={column.key}
                    style={column.width ? { width: column.width } : {}}
                    class={column.align ? `text-${column.align}` : ""}
                  >
                    {column.title}
                  </th>
                ))}
              </tr>
            </thead>
            <tbody>
              {Array.from({ length: 5 }, (_, index) => (
                <tr key={index}>
                  {safeColumns.map((column) => (
                    <td key={column.key}>
                      <div class="skeleton h-4 w-full"></div>
                    </td>
                  ))}
                </tr>
              ))}
            </tbody>
          </table>
        </div>
        {renderPaginationInfo()}
      </div>
    );
  }

  return (
    <div class="space-y-4">
      {renderControls()}
      <div class="overflow-x-auto">
        <table class={tableClasses} id={id} {...props}>
          <thead>
            <tr>
              {safeColumns.map((column) => (
                <th
                  key={column.key}
                  style={column.width ? { width: column.width } : {}}
                  class={[
                    column.align ? `text-${column.align}` : "",
                    sortable && column.sortable ? "cursor-pointer hover:bg-base-200" : "",
                  ].filter(Boolean).join(" ")}
                  onClick={() => handleSort(column.key)}
                >
                  <div class="flex items-center gap-2">
                    <span>{column.title}</span>
                    {sortable && column.sortable && getSortIcon(column.key)}
                  </div>
                </th>
              ))}
            </tr>
          </thead>
          <tbody>
            {safeData.length === 0
              ? (
                <tr>
                  <td colSpan={safeColumns.length} class="text-center py-8 text-base-content/70">
                    {emptyMessage}
                  </td>
                </tr>
              )
              : (
                displayData.map((row, index) => (
                  <tr key={index}>
                    {safeColumns.map((column) => (
                      <td
                        key={column.key}
                        class={column.align ? `text-${column.align}` : ""}
                      >
                        {column.render ? column.render(row[column.key], row) : String(row[column.key] ?? "")}
                      </td>
                    ))}
                  </tr>
                ))
              )}
          </tbody>
        </table>
      </div>
      {renderPaginationInfo()}
    </div>
  );
}

export function PaginatedTable({
  class: className = "",
  columns = [],
  data = [],
  zebra = false,
  compact = false,
  hover = false,
  pinRows = false,
  pinCols = false,
  emptyMessage = "No data available",
  initialItemsPerPage = 10,
  itemsPerPageOptions = [5, 10, 20, 50],
  showStats = true,
  showControls = true,
  allowSorting = true,
  allowItemsPerPageChange = true,
  id,
  ...props
}: PaginatedTableProps) {
  const [currentPage, setCurrentPage] = useState(1);
  const [itemsPerPage, setItemsPerPage] = useState(initialItemsPerPage);
  // Ensure columns and data are arrays
  const safeColumns = Array.isArray(columns) ? columns : [];
  const safeData = Array.isArray(data) ? data : [];
  
  const [sortColumn, setSortColumn] = useState(safeColumns[0]?.key || "");
  const [sortDirection, setSortDirection] = useState<"asc" | "desc">("asc");

  // Sort data
  const sortedData = allowSorting
    ? [...safeData].sort((a, b) => {
      const aValue = a[sortColumn];
      const bValue = b[sortColumn];
      
      // Convert to strings for comparison
      const aStr = String(aValue ?? "");
      const bStr = String(bValue ?? "");
      
      const comparison = aStr < bStr ? -1 : aStr > bStr ? 1 : 0;
      return sortDirection === "asc" ? comparison : -comparison;
    })
    : safeData;

  const handleSort = (column: string, direction: "asc" | "desc") => {
    setSortColumn(column);
    setSortDirection(direction);
    setCurrentPage(1); // Reset to first page when sorting
  };

  const handlePageChange = (page: number) => {
    setCurrentPage(page);
  };

  const handleItemsPerPageChange = (newItemsPerPage: number) => {
    setItemsPerPage(newItemsPerPage);
    setCurrentPage(1); // Reset to first page when changing page size
  };

  return (
    <Table
      class={className}
      columns={safeColumns}
      data={sortedData}
      zebra={zebra}
      compact={compact}
      hover={hover}
      pinRows={pinRows}
      pinCols={pinCols}
      emptyMessage={emptyMessage}
      paginated
      currentPage={currentPage}
      itemsPerPage={itemsPerPage}
      totalItems={safeData.length}
      onPageChange={handlePageChange}
      onItemsPerPageChange={handleItemsPerPageChange}
      showControls={showControls}
      allowItemsPerPageChange={allowItemsPerPageChange}
      showStats={showStats}
      itemsPerPageOptions={itemsPerPageOptions}
      sortable={allowSorting}
      sortColumn={sortColumn}
      sortDirection={sortDirection}
      onSort={handleSort}
      id={id}
      {...props}
    />
  );
}
