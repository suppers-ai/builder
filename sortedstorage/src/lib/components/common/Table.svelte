<script lang="ts">
	import { ChevronUp, ChevronDown } from 'lucide-svelte';
	
	export let columns: {
		key: string;
		label: string;
		sortable?: boolean;
		width?: string;
		align?: 'left' | 'center' | 'right';
	}[] = [];
	
	export let data: any[] = [];
	export let sortBy = '';
	export let sortOrder: 'asc' | 'desc' = 'asc';
	export let selectable = false;
	export let selectedRows: any[] = [];
	export let loading = false;
	export let emptyMessage = 'No data available';
	
	let className = '';
	export { className as class };
	
	function handleSort(column: string) {
		if (sortBy === column) {
			sortOrder = sortOrder === 'asc' ? 'desc' : 'asc';
		} else {
			sortBy = column;
			sortOrder = 'asc';
		}
	}
	
	function toggleSelection(row: any) {
		if (selectedRows.includes(row)) {
			selectedRows = selectedRows.filter(r => r !== row);
		} else {
			selectedRows = [...selectedRows, row];
		}
	}
	
	function toggleSelectAll() {
		if (selectedRows.length === data.length) {
			selectedRows = [];
		} else {
			selectedRows = [...data];
		}
	}
	
	$: allSelected = data.length > 0 && selectedRows.length === data.length;
	$: someSelected = selectedRows.length > 0 && selectedRows.length < data.length;
</script>

<div class="overflow-x-auto {className}">
	<table class="w-full">
		<thead class="bg-gray-50 dark:bg-gray-800 border-b border-gray-200 dark:border-gray-700">
			<tr>
				{#if selectable}
					<th class="px-4 py-3 text-left">
						<input
							type="checkbox"
							checked={allSelected}
							indeterminate={someSelected}
							on:change={toggleSelectAll}
							class="rounded border-gray-300 dark:border-gray-600"
						/>
					</th>
				{/if}
				{#each columns as column}
					<th 
						class="px-4 py-3 text-{column.align || 'left'} text-sm font-medium text-gray-700 dark:text-gray-300"
						style={column.width ? `width: ${column.width}` : ''}
					>
						{#if column.sortable}
							<button
								on:click={() => handleSort(column.key)}
								class="flex items-center gap-1 hover:text-primary-600 transition-colors"
							>
								{column.label}
								<span class="flex flex-col">
									<ChevronUp 
										class="w-3 h-3 -mb-1 {sortBy === column.key && sortOrder === 'asc' ? 'text-primary-600' : 'text-gray-400'}"
									/>
									<ChevronDown 
										class="w-3 h-3 {sortBy === column.key && sortOrder === 'desc' ? 'text-primary-600' : 'text-gray-400'}"
									/>
								</span>
							</button>
						{:else}
							{column.label}
						{/if}
					</th>
				{/each}
			</tr>
		</thead>
		<tbody class="divide-y divide-gray-200 dark:divide-gray-700">
			{#if loading}
				<tr>
					<td colspan={columns.length + (selectable ? 1 : 0)} class="px-4 py-8 text-center">
						<div class="animate-spin w-8 h-8 border-2 border-primary-500 border-t-transparent rounded-full mx-auto"></div>
						<p class="text-gray-500 mt-2">Loading...</p>
					</td>
				</tr>
			{:else if data.length === 0}
				<tr>
					<td colspan={columns.length + (selectable ? 1 : 0)} class="px-4 py-8 text-center text-gray-500">
						{emptyMessage}
					</td>
				</tr>
			{:else}
				{#each data as row}
					<tr 
						class="hover:bg-gray-50 dark:hover:bg-gray-800 transition-colors"
						class:bg-primary-50={selectedRows.includes(row)}
						class:dark:bg-primary-900={selectedRows.includes(row)}
					>
						{#if selectable}
							<td class="px-4 py-3">
								<input
									type="checkbox"
									checked={selectedRows.includes(row)}
									on:change={() => toggleSelection(row)}
									class="rounded border-gray-300 dark:border-gray-600"
								/>
							</td>
						{/if}
						{#each columns as column}
							<td class="px-4 py-3 text-{column.align || 'left'}">
								<slot name="cell" {column} {row} value={row[column.key]}>
									{row[column.key]}
								</slot>
							</td>
						{/each}
					</tr>
				{/each}
			{/if}
		</tbody>
	</table>
</div>