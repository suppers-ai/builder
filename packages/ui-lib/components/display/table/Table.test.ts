import { assertEquals, assertStringIncludes } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import { renderToString } from "preact-render-to-string";
import { Table } from "./Table.tsx";

Deno.test("Table - basic rendering", () => {
  const html = renderToString(Table({
    columns: [{ key: "name", title: "Name" }],
    data: [{ name: "Table content" }],
  }));
  assertStringIncludes(html, "table");
  assertStringIncludes(html, "Table content");
});

Deno.test("Table - with headers and data", () => {
  const columns = [
    { key: "name", title: "Name" },
    { key: "age", title: "Age" },
    { key: "email", title: "Email" },
  ];
  const data = [
    { name: "John Doe", age: "30", email: "john@example.com" },
    { name: "Jane Smith", age: "25", email: "jane@example.com" },
  ];

  const html = renderToString(Table({
    columns,
    data,
  }));
  assertStringIncludes(html, "table");
  assertStringIncludes(html, "Name");
  assertStringIncludes(html, "Age");
  assertStringIncludes(html, "Email");
  assertStringIncludes(html, "John Doe");
  assertStringIncludes(html, "jane@example.com");
});

Deno.test("Table - zebra striping", () => {
  const html = renderToString(Table({
    columns: [{ key: "name", title: "Name" }],
    data: [{ name: "Content" }],
    zebra: true,
  }));
  assertStringIncludes(html, "table-zebra");
});

Deno.test("Table - compact size", () => {
  const html = renderToString(Table({
    columns: [{ key: "name", title: "Name" }],
    data: [{ name: "Content" }],
    compact: true,
  }));
  assertStringIncludes(html, "table-xs");
});

Deno.test("Table - hover behavior", () => {
  const html = renderToString(Table({
    columns: [{ key: "name", title: "Name" }],
    data: [{ name: "Content" }],
    hover: true,
  }));
  assertStringIncludes(html, "table-hover");
});

Deno.test("Table - loading state", () => {
  const html = renderToString(Table({
    columns: [{ key: "name", title: "Name" }],
    data: [],
    loading: true,
  }));
  assertStringIncludes(html, "skeleton");
});

Deno.test("Table - empty state", () => {
  const html = renderToString(Table({
    columns: [{ key: "name", title: "Name" }],
    data: [],
    emptyMessage: "No data",
  }));
  assertStringIncludes(html, "No data");
});

Deno.test("Table - with pin rows", () => {
  const html = renderToString(Table({
    columns: [{ key: "name", title: "Name" }],
    data: [{ name: "Content" }],
    pinRows: true,
  }));
  assertStringIncludes(html, "table-pin-rows");
});

Deno.test("Table - with pin columns", () => {
  const html = renderToString(Table({
    columns: [{ key: "name", title: "Name" }],
    data: [{ name: "Content" }],
    pinCols: true,
  }));
  assertStringIncludes(html, "table-pin-cols");
});

Deno.test("Table - with custom class", () => {
  const html = renderToString(Table({
    columns: [{ key: "name", title: "Name" }],
    data: [{ name: "Content" }],
    class: "custom-table",
  }));
  assertStringIncludes(html, "custom-table");
});

Deno.test("Table - responsive", () => {
  const html = renderToString(Table({
    columns: [{ key: "name", title: "Name" }],
    data: [{ name: "Content" }],
  }));
  assertStringIncludes(html, "overflow-x-auto");
});

Deno.test("Table - HTML snapshot", async (t) => {
  const columns = [
    { key: "product", title: "Product" },
    { key: "price", title: "Price" },
    { key: "stock", title: "Stock" },
  ];
  const data = [
    { product: "Widget A", price: "$10.99", stock: "15" },
    { product: "Widget B", price: "$25.50", stock: "8" },
    { product: "Widget C", price: "$5.00", stock: "32" },
  ];

  const html = renderToString(Table({
    columns,
    data,
    zebra: true,
  }));
  await assertSnapshot(t, html);
});

Deno.test("Table - HTML snapshot compact", async (t) => {
  const html = renderToString(Table({
    columns: [
      { key: "id", title: "ID" },
      { key: "status", title: "Status" },
    ],
    data: [
      { id: "001", status: "Active" },
      { id: "002", status: "Inactive" },
    ],
    compact: true,
    pinRows: true,
  }));
  await assertSnapshot(t, html);
});
