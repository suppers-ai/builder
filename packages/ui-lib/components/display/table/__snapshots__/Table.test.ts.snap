export const snapshot = {};

snapshot[`Table - HTML snapshot 1`] = `'<div class="space-y-4"><div class="overflow-x-auto"><table class="table table-zebra"><thead><tr><th class><div class="flex items-center gap-2"><span>Product</span></div></th><th class><div class="flex items-center gap-2"><span>Price</span></div></th><th class><div class="flex items-center gap-2"><span>Stock</span></div></th></tr></thead><tbody><tr><td class>Widget A</td><td class>$10.99</td><td class>15</td></tr><tr><td class>Widget B</td><td class>$25.50</td><td class>8</td></tr><tr><td class>Widget C</td><td class>$5.00</td><td class>32</td></tr></tbody></table></div></div>'`;

snapshot[`Table - HTML snapshot compact 1`] = `'<div class="space-y-4"><div class="overflow-x-auto"><table class="table table-xs table-pin-rows"><thead><tr><th class><div class="flex items-center gap-2"><span>ID</span></div></th><th class><div class="flex items-center gap-2"><span>Status</span></div></th></tr></thead><tbody><tr><td class>001</td><td class>Active</td></tr><tr><td class>002</td><td class>Inactive</td></tr></tbody></table></div></div>'`;
